### PR TITLE
treewide: replace boost::find_if with std::ranges::find_if

### PR DESCRIPTION
**Pull request by tchaikov**
now that we are allowed to use C++23. we now have the luxury of using `std::ranges::find_if`.

in this change, we:

- replace `boost::find_if` with `std::ranges::find_if`
- remove all `#include <boost/range/algorithm/find_if.hpp>`

to reduce the dependency to boost for better maintainability, and leverage standard library features for better long-term support.

this change is part of our ongoing effort to modernize our codebase and reduce external dependencies where possible.

---

it's a cleanup, hence no need to backport.

### CODE REVIEW

File: counters.hh [hunk 1]
```diff
@@ -253,23 +252,31 @@ protected:
...
-        using reference = basic_counter_shard_view<is_mutable>&;
+        using reference = std::conditional_t<is_mutable == mutable_view::yes,
+            basic_counter_shard_view<is_mutable>&,
+            const basic_counter_shard_view<is_mutable>&>;
     private:
         managed_bytes_basic_view<is_mutable> _current;
         basic_counter_shard_view<is_mutable> _current_view;
         size_t _pos = 0;
     public:
+        shard_iterator() noexcept = default;
```
Comment (regevran) [inline counters.hh:267]: With this change `_current_view` is _not_ constructed with `_current`.
It is not the same as-if before the change:
```cpp
shard_iterator(managed_bytes_basic_view<mutable_view::yes>(), 0) // or mutable_view::no
```

Why do we need default empty constructor for?

File: tools/scylla-sstable.cc [hunk 2]
```diff
@@ -996,7 +996,7 @@ void scrub_operation(schema_ptr schema, reader_permit permit, const std::vector<
             throw std::invalid_argument("missing mandatory command-line argument --scrub-mode");
         }
         const auto mode_name = vm["scrub-mode"].as<std::string>();
-        auto mode_it = boost::find_if(scrub_modes, [&mode_name] (const std::pair<std::string, compaction_type_options::scrub::mode>& v) {
+        auto mode_it = std::ranges::find_if(scrub_modes, [&mode_name] (const std::pair<std::string, compaction_type_options::scrub::mode>& v) {
```
Comment (regevran) [inline tools/scylla-sstable.cc:999]: Nitpick:
Here and in a few other places you replaced `boost::find_if` with `std::ranges::find_if`.
You could replace instead with `std::find_if` (no ranges involved) - but I can see the benefit of using one algorithm in all places (especially when it consumes the other).

**Review by regevran — CHANGES_REQUESTED** [review]
Please replace the default constructor with one that set the internal members correctly.

**Comment by avikivity** [comment]
> Please replace the default constructor with one that set the internal members correctly.

It's unnecessary, just like a pointer's default constructor doesn't do anything. One can't do anything with a default-constructed iterator except assign to it.

File: counters.hh [hunk 3]
```diff
@@ -253,23 +252,31 @@ protected:
 private:
     class shard_iterator {
     public:
-        using iterator_category = std::input_iterator_tag;
+        using iterator_category = std::forward_iterator_tag;
```
Comment (avikivity) [inline counters.hh:255]: Please set iterator_concept too.

File: counters.hh (hunk 1 above)
Comment (avikivity) [reply counters.hh:267]: Iterator concepts require it, but it doesn't need to do anything.

File: counters.hh [hunk 4]
```diff
@@ -253,23 +252,31 @@ protected:
...
         basic_counter_shard_view<is_mutable> _current_view;
         size_t _pos = 0;
     public:
+        shard_iterator() noexcept = default;
         shard_iterator(managed_bytes_basic_view<is_mutable> v, size_t offset) noexcept
             : _current(v), _current_view(_current), _pos(offset) { }
 
-        basic_counter_shard_view<is_mutable>& operator*() noexcept {
+        reference operator*() const noexcept requires (is_mutable == mutable_view::no) {
             return _current_view;
```
Comment (avikivity) [inline counters.hh:272]: If you return a reference to the iterator itself, then it's not a forward_iterator.

**Comment by regevran** [comment]
> > Please replace the default constructor with one that set the internal members correctly.
> 
> It's unnecessary, just like a pointer's default constructor doesn't do anything. One can't do anything with a default-constructed iterator except assign to it.

Looking at the code it seems that indeed no harm would be made (as `basic_counter_shard_view` is default constructable, it does not need to be constructed with a `managed_bytes_basic_view`).
However, with the following two snippets one would expect to get the same result, but with default constructed `shard_iterator` it is not the same, the results are different:
```cpp
// 1
shard_iterator it(managed_bytes_basic_view<mutable_view::yes>(), 0);
// 2
shard_iterator it;
```

File: tools/scylla-sstable.cc (hunk 2 above)
Comment (tchaikov) [reply tools/scylla-sstable.cc:999]: i don't see why it's preferred to use `std::find_if` here.

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help review this change?

**Comment by tchaikov** [comment]
v3:

- rebased against master HEAD

**Comment by tchaikov** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13235/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13235/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_compaction_test 🔹 boost/sstable_datafile_test ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/275/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13235/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1752/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4837/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12854/)
> #### Failed Tests (32/16179):
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_fragments_monotonic_cg0.3.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_fragments_monotonic_cg1.2.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_fragments_monotonic_cg1.3.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_read_back_cg0.3.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_read_back_cg1.2.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_read_back_cg1.3.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_reader_conversion_cg0.3.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_reader_conversion_cg1.2.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_reader_conversion_cg1.3.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13235/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_reverse_fragments_monotonic_cg0.2.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
> 
> #### Build Details:
> 
>     * Duration: 2 hr 22 min
> 
>     * Builder: spider3.cloudius-systems.com

```
12:30:29  [482/29790]  boost    debug  [ PASS ] boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_reader_conversion_cg0.2 110.58s
12:30:49 
 Cannot contact spider3.cloudius-systems.com: java.lang.InterruptedException
```

i believe this is an intermediate infra issue.

**Review by denesb — APPROVED** [review]

**Pull request merged**